dummyfs/object.c

```diff
@@ -22,127 +22,149 @@
 		free(r);
-		mutexUnlock(dummyfs_common.mutex);
-		mutexUnlock(olock);
+		mutexUnlock(ctx->mutex);
+		mutexUnlock(ctx->olock);
 		return NULL;
 	}
 
-	if (dummyfs_incsz(sizeof(dummyfs_object_t)) != EOK) {
+	if (dummyfs_incsz(ctx, sizeof(dummyfs_object_t)) != EOK) {
 		free(r);
-		mutexUnlock(dummyfs_common.mutex);
-		mutexUnlock(olock);
+		mutexUnlock(ctx->mutex);
+		mutexUnlock(ctx->olock);
 		return NULL;
 	}
 
-	mutexUnlock(dummyfs_common.mutex);
+	mutexUnlock(ctx->mutex);
 
 	r->oid.id = id;
 	r->refs = 1;
 	r->mode = 0;
 	r->nlink = 0;
 
-	mutexUnlock(olock);
+	mutexUnlock(ctx->olock);
 
 	return r;
 }
 
 
-void object_lock(dummyfs_object_t *o)
+void object_lock(dummyfs_t *ctx, dummyfs_object_t *o)
 {
-	mutexLock(dummyfs_common.mutex);
+	mutexLock(ctx->mutex);
 }
 
 
-void object_unlock(dummyfs_object_t *o)
+void object_unlock(dummyfs_t *ctx, dummyfs_object_t *o)
 {
-	mutexUnlock(dummyfs_common.mutex);
+	mutexUnlock(ctx->mutex);
 }
 
 
-int object_remove(dummyfs_object_t *o)
+int object_remove(dummyfs_t *ctx, dummyfs_object_t *o)
 {
 	if (o->nlink || o->refs)
 		return -EBUSY;
 
-	idtree_remove(&dummytree, &o->node);
+	idtree_remove(&ctx->dummytree, &o->node);
 	return EOK;
 }
 
 
-dummyfs_object_t *object_get_unlocked(unsigned int id)
+dummyfs_object_t *object_get_unlocked(dummyfs_t *ctx, unsigned int id)
 {
 	dummyfs_object_t *o;
 
-	o = dummy_node2obj(idtree_find(&dummytree, id));
+	o = dummy_node2obj(idtree_find(&ctx->dummytree, id));
 
 	return o;
 }
 
 
-dummyfs_object_t *object_get(unsigned int id)
+dummyfs_object_t *object_get(dummyfs_t *ctx, unsigned int id)
 {
 	dummyfs_object_t *o;
 
-	mutexLock(olock);
-	if ((o = object_get_unlocked(id)) != NULL)
+	mutexLock(ctx->olock);
+	if ((o = object_get_unlocked(ctx, id)) != NULL)
 		o->refs++;
-	mutexUnlock(olock);
+	mutexUnlock(ctx->olock);
 
 	return o;
 }
 
 
-void object_put(dummyfs_object_t *o)
+void object_put(dummyfs_t *ctx, dummyfs_object_t *o)
 {
-	mutexLock(olock);
+	mutexLock(ctx->olock);
 	if (o != NULL && o->refs) {
 		o->refs--;
 
 		if (!o->refs && S_ISDIR(o->mode) && o->dirty) {
-			object_lock(o);
-			dir_clean(o);
-			object_unlock(o);
+			object_lock(ctx, o);
+			dir_clean(ctx, o);
+			object_unlock(ctx, o);
 		}
 
 		if (!o->refs && !o->nlink)
-			dummyfs_destroy(&o->oid);
+			dummyfs_destroy(ctx, &o->oid);
 	}
-	mutexUnlock(olock);
+	mutexUnlock(ctx->olock);
 	return;
 }
 
 
-void object_init(void)
+int object_init(dummyfs_t *ctx)
 {
-	idtree_init(&dummytree);
-	mutexCreate(&olock);
+	if (mutexCreate(&ctx->olock) != EOK)
+		return -ENOMEM;
+
+	idtree_init(&ctx->dummytree);
+
+	return EOK;
+}
+
+
+void object_cleanup(dummyfs_t *ctx)
+{
+	dummyfs_object_t *o;
+	rbnode_t *n;
+
+	mutexLock(ctx->olock);
+	/* Forcibly destroy all objects, ignore refs and links */
+	while ((n = lib_rbMinimum(ctx->dummytree.root)) != NULL) {
+		o = dummy_node2obj(n);
+		if (S_ISDIR(o->mode))
+			dir_clean(ctx, o);
+		idtree_remove(&ctx->dummytree, &o->node);
+		dummyfs_destroy(ctx, &o->oid);
```

Comment:
Hmmm, at least ctx->mutex is taken inside. I think that I've seen ctx->mutex being taken before ctx->olock - this is deadlock scenario

Reply:
It seems, that `dummyfs_destroy()` should always be called under `ctx->olock`, but it's not used correctly in other places. Here I don't see any deadlock scenarios, but I see unsynchronized accesses elsewhere.

Reply:
After a thorough analysis I'd say, that whole synchronization should be fixed. It works fine at the moment, because we have only one thread accessing the filesystem. Once we add more threads we can get into trouble. Apart from the mentioned `dummyfs_destroy()`, we have  e.g. `dummyfs_incsz()` or `dummyfs_decsz()`, which are called in different contexts, but most of the time, no mutex is taken.

I propose, that we leave it as it is and maybe create an issue, that we need a bigger refactoring of dummyfs if we want libdummyfs to be thread-safe, or we could accept the fact that it is not thread-safe and remove the mutexes to avoid wasting CPU time.

Reply:
I agree, these fixes are outside of the scope of this PR